`TestKubernetesLicenseEnforcement ` flakiness

## Failure

#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/4834472626/jobs/8615732335

#### Relevant snippet

```
===================================================
OUTPUT github.com/gravitational/teleport/lib/kube/proxy.TestKubernetesLicenseEnforcement/kubernetes_agent_is_licensed
===================================================
make: *** [test-go-unit] Error 1
=== RUN   TestKubernetesLicenseEnforcement/kubernetes_agent_is_licensed
=== PAUSE TestKubernetesLicenseEnforcement/kubernetes_agent_is_licensed
=== CONT  TestKubernetesLicenseEnforcement/kubernetes_agent_is_licensed
{"caller":"proxy/auth.go:140","cluster":"foo","component":null,"level":"debug","message":"Checking Kubernetes impersonation permissions.","timestamp":"2023-04-28T20:28:35Z"}
{"caller":"proxy/auth.go:140","cluster":"teleport-cluster","component":null,"level":"debug","message":"Checking Kubernetes impersonation permissions.","timestamp":"2023-04-28T20:28:35Z"}
{"caller":"proxy/auth.go:140","cluster":"baz","component":null,"level":"debug","message":"Checking Kubernetes impersonation permissions.","timestamp":"2023-04-28T20:28:35Z"}
{"caller":"proxy/auth.go:151","cluster":"teleport-cluster","component":null,"level":"debug","message":"Have all necessary Kubernetes impersonation permissions.","timestamp":"2023-04-28T20:28:35Z"}
{"caller":"proxy/auth.go:176","cluster":"teleport-cluster","component":null,"level":"debug","message":"Initialized Kubernetes credentials","timestamp":"2023-04-28T20:28:35Z"}
{"caller":"proxy/auth.go:73","component":null,"kubeClusterName":"","kubeconfigPath":"/tmp/TestGetKubeCreds4273838333/001/kubeconfig","level":"debug","message":"Reading Kubernetes details.","serviceType":"kube_proxy","timestamp":"2023-04-28T20:28:35Z"}
    testing.go:1446: race detected during execution of test
--- FAIL: TestKubernetesLicenseEnforcement/kubernetes_agent_is_licensed (0.92s)
time="2023-04-28T20:28:36Z" level=warning msg="Mux serve failed." error="listener is closed\n\taccept tcp 127.0.0.1:43395: use of closed network connection" trace.component=auth
time="2023-04-28T20:28:36Z" level=warning msg="Mux serve failed." error="listener is closed\n\taccept tcp 127.0.0.1:46393: use of closed network connection" trace.component=auth
===================================================
OUTPUT github.com/gravitational/teleport/lib/kube/proxy.TestKubernetesLicenseEnforcement/kubernetes_isn't_licensed
===================================================
=== RUN   TestKubernetesLicenseEnforcement/kubernetes_isn't_licensed
=== PAUSE TestKubernetesLicenseEnforcement/kubernetes_isn't_licensed
{"caller":"proxy/auth.go:73","component":null,"kubeClusterName":"","kubeconfigPath":"/tmp/TestGetKubeCreds4273838333/001/kubeconfig","level":"debug","message":"Reading Kubernetes details.","serviceType":"kube_service","timestamp":"2023-04-28T20:28:35Z"}
=== CONT  TestKubernetesLicenseEnforcement/kubernetes_isn't_licensed
{"caller":"proxy/auth.go:73","component":null,"kubeClusterName":"","kubeconfigPath":"","level":"debug","message":"Reading Kubernetes details.","serviceType":"kube_proxy","timestamp":"2023-04-28T20:28:35Z"}
    testing.go:1446: race detected during execution of test
--- FAIL: TestKubernetesLicenseEnforcement/kubernetes_isn't_licensed (0.92s)
```

--- a/lib/kube/proxy/forwarder_test.go
+++ b/lib/kube/proxy/forwarder_test.go
@@ -1903,7 +1903,7 @@ func TestKubernetesLicenseEnforcement(t *testing.T) {
 				kubeCluster,
 			)
 
-			_, err = client.CoreV1().Pods(metav1.NamespaceDefault).List(context.Background(), metav1.ListOptions{})
+			_, err := client.CoreV1().Pods(metav1.NamespaceDefault).List(context.Background(), metav1.ListOptions{})
 			tt.assertErrFunc(t, err)
 		})
 	}

--- a/lib/kube/proxy/utils_testing.go
+++ b/lib/kube/proxy/utils_testing.go
@@ -56,17 +56,19 @@ import (
 )
 
 type TestContext struct {
-	HostID      string
-	ClusterName string
-	TLSServer   *auth.TestTLSServer
-	AuthServer  *auth.Server
-	AuthClient  *auth.Client
-	Authz       authz.Authorizer
-	KubeServer  *TLSServer
-	Emitter     *eventstest.ChannelEmitter
-	Context     context.Context
-	listener    net.Listener
-	cancel      context.CancelFunc
+	HostID          string
+	ClusterName     string
+	TLSServer       *auth.TestTLSServer
+	AuthServer      *auth.Server
+	AuthClient      *auth.Client
+	Authz           authz.Authorizer
+	KubeServer      *TLSServer
+	Emitter         *eventstest.ChannelEmitter
+	Context         context.Context
+	listener        net.Listener
+	cancel          context.CancelFunc
+	heartbeatCtx    context.Context
+	heartbeatCancel context.CancelFunc
 }
 
 // KubeClusterConfig defines the cluster to be created
@@ -89,11 +91,14 @@ type TestConfig struct {
 // SetupTestContext creates a kube service with clusters configured.
 func SetupTestContext(ctx context.Context, t *testing.T, cfg TestConfig) *TestContext {
 	ctx, cancel := context.WithCancel(ctx)
+	heartbeatCtx, heartbeatCancel := context.WithCancel(ctx)
 	testCtx := &TestContext{
-		ClusterName: "root.example.com",
-		HostID:      uuid.New().String(),
-		Context:     ctx,
-		cancel:      cancel,
+		ClusterName:     "root.example.com",
+		HostID:          uuid.New().String(),
+		Context:         ctx,
+		cancel:          cancel,
+		heartbeatCtx:    heartbeatCtx,
+		heartbeatCancel: heartbeatCancel,
 	}
 	t.Cleanup(func() { testCtx.Close() })
 
@@ -223,11 +228,17 @@ func SetupTestContext(ctx context.Context, t *testing.T, cfg TestConfig) *TestCo
 		// this is used to make sure that heartbeat started and the clusters
 		// are registered in the auth server
 		OnHeartbeat: func(err error) {
-			// ignore not found errors because although the heartbeat is called before
-			// the close does not wait for the resource cleanup to finish.
-			if trace.IsNotFound(err) {
-				return
+			select {
+			case <-heartbeatCtx.Done():
+				// ignore not found errors because although the heartbeat is called before
+				// the close does not wait for the resource cleanup to finish.
+				if trace.IsNotFound(err) {
+					return
+				}
+			default:
+
 			}
+
 			assert.NoError(t, err)
 			select {
 			case heartbeatsWaitChannel <- struct{}{}:
@@ -269,6 +280,9 @@ func (c *TestContext) startKubeService(t *testing.T) {
 
 // Close closes resources associated with the test context.
 func (c *TestContext) Close() error {
+	// cancel the heartbeat context to stop validating the heartbeat not found
+	// errors when deprovisioning.
+	c.heartbeatCancel()
 	// kubeServer closes the listener
 	err := c.KubeServer.Close()
 	authCErr := c.AuthClient.Close()

--- a/lib/kube/proxy/utils_testing.go
+++ b/lib/kube/proxy/utils_testing.go
@@ -223,7 +223,12 @@ func SetupTestContext(ctx context.Context, t *testing.T, cfg TestConfig) *TestCo
 		// this is used to make sure that heartbeat started and the clusters
 		// are registered in the auth server
 		OnHeartbeat: func(err error) {
-			require.NoError(t, err)
+			// ignore not found errors because although the heartbeat is called before
+			// the close does not wait for the resource cleanup to finish.
+			if trace.IsNotFound(err) {
+				return
+			}
+			assert.NoError(t, err)
 			select {
 			case heartbeatsWaitChannel <- struct{}{}:
 			default:

